Sources/NIO/SelectorKqueue.swift

```diff
@@ -0,0 +1,356 @@
+    init(selectorEventSet: SelectorEventSet) {
+        var kqueueFilterSet: KQueueEventFilterSet = .init(rawValue: 0)
+        if selectorEventSet.contains(.read) {
+            kqueueFilterSet.formUnion(.read)
+        }
+
+        if selectorEventSet.contains(.write) {
+            kqueueFilterSet.formUnion(.write)
+        }
+
+        if isEarlyEOFDeliveryWorkingOnThisOS && selectorEventSet.contains(.readEOF) {
+            kqueueFilterSet.formUnion(.except)
+        }
+        self = kqueueFilterSet
+    }
+
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    /// Calculate the kqueue filter changes that are necessary to transition from `previousKQueueFilterSet` to `self`.
+    /// The `body` closure is then called with the changes necessary expressed as a number of `kevent`.
+    ///
+    /// - parameters:
+    ///    - previousKQueueFilterSet: The previous filter set that is currently registered with kqueue.
+    ///    - fileDescriptor: The file descriptor the `kevent`s should be generated to.
+    ///    - body: The closure that will then apply the change set.
+    func calculateKQueueFilterSetChanges(previousKQueueFilterSet: KQueueEventFilterSet,
+                                         fileDescriptor: CInt,
+                                         _ body: (UnsafeMutableBufferPointer<kevent>) throws -> Void) rethrows {
+        // we only use three filters (EVFILT_READ, EVFILT_WRITE and EVFILT_EXCEPT) so the number of changes would be 3.
+        var keventsHopefullyOnStack = (kevent(), kevent(), kevent())
+        try withUnsafeMutableBytes(of: &keventsHopefullyOnStack) { rawPtr in
+            assert(MemoryLayout<kevent>.size * 3 == rawPtr.count)
+            let keventBuffer = rawPtr.baseAddress!.bindMemory(to: kevent.self, capacity: 3)
+
+            let differences = previousKQueueFilterSet.symmetricDifference(self) // contains all the events that need a change (either need to be added or removed)
+
+            func calculateKQueueChange(event: KQueueEventFilterSet) -> UInt16? {
+                guard differences.contains(event) else {
+                    return nil
+                }
+                return UInt16(self.contains(event) ? EV_ADD : EV_DELETE)
+            }
+
+            var index: Int = 0
+            for (event, filter) in [(KQueueEventFilterSet.read, EVFILT_READ), (.write, EVFILT_WRITE), (.except, EVFILT_EXCEPT)] {
+                if let flags = calculateKQueueChange(event: event) {
+                    keventBuffer[index].setEvent(fileDescriptor: fileDescriptor, filter: filter, flags: flags)
+                    index += 1
+                }
+            }
+            try body(UnsafeMutableBufferPointer(start: keventBuffer, count: index))
+        }
+    }
+    #endif
+}
+
+/* this is deliberately not thread-safe, only the wakeup() function may be called unprotectedly */
+final internal class KqueueSelector<R: Registration>: Selector<R> {
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```

Comment:
here as well, can we make this one huge `#if os...` around the whole file?

Reply:
Fixed in e19e685.